### PR TITLE
Add characteristics measures

### PR DESCRIPTION
**Pull request by julienlancelot**
- Add rating measures
- Add effort to grade A measures

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/QualityModelMeasuresVisitor.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+  private void addEffortToRatingAMeasure(Component component, Metric effortMetric, Metric effortToRatingAMetric, Path<QualityModelCounter> path) {
+    long developmentCostValue = path.current().devCosts;
+    Optional<Measure> effortMeasure = measureRepository.getRawMeasure(component, effortMetric);
+    long effort = effortMeasure.isPresent() ? effortMeasure.get().getIntValue() : 0L;
+    long upperGradeCost = ((Double) (ratingGrid.getGradeLowerBound(Rating.B) * developmentCostValue)).longValue();
+    long effortToRatingA = upperGradeCost < effort ? (effort - upperGradeCost) : 0L;
+    measureRepository.add(component, effortToRatingAMetric, Measure.newMeasureBuilder().create(effortToRatingA));
+  }
+
+  private void addToParent(Path<QualityModelCounter> path) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/QualityModelMeasuresVisitor.java]: MAJOR Make "addToParent" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 17 issues:
* MAJOR 6 major
* INFO 11 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2164) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2170) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2182) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2188) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2202) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2208) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2237) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2243) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2256) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonarqube/blob/5541b4c07acb6832278cd8b3f9a16e14e6bfa29b/sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java#L2262) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)
* ... 6 more

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define the locale to be used in this String operation.](https://github.com/SonarSource/sonarqube/blob/dddebcef56e11021bad170d37e27c535232a2f6c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L166) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1449)
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/dddebcef56e11021bad170d37e27c535232a2f6c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/dddebcef56e11021bad170d37e27c535232a2f6c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L260) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)
* MAJOR [Make "computeDensity" a "static" method.](https://github.com/SonarSource/sonarqube/blob/dddebcef56e11021bad170d37e27c535232a2f6c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L128) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)
* MAJOR [Define the locale to be used in this String operation.](https://github.com/SonarSource/sonarqube/blob/dddebcef56e11021bad170d37e27c535232a2f6c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/RatingSettings.java#L81) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1449)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define the locale to be used in this String operation.](https://github.com/SonarSource/sonarqube/blob/d1336d888ab60c6ad8e863d263f05a2a3169028c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L166) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1449)
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/d1336d888ab60c6ad8e863d263f05a2a3169028c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/d1336d888ab60c6ad8e863d263f05a2a3169028c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L260) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)
* MAJOR [Define the locale to be used in this String operation.](https://github.com/SonarSource/sonarqube/blob/d1336d888ab60c6ad8e863d263f05a2a3169028c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/RatingSettings.java#L81) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1449)

**Comment by teryk** [comment]
`effort_to_reach_maintainability_rating_a0`does not seem to work properly, the rest is good though!
the code review is high level and it looks good

**Comment by julienlancelot** [comment]
Thanks for the feedback, indeed the effort_to_reach_maintainability_rating_a was based on a wrong metric ! I've used 'code_smells' (which is the number of issues of type CODE SMELL) when I should have used 'sqale_index' (which is the amount of effort of issues of type CODE SMELL)...
It's now fixed, please test again @teryk.

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 2]
```diff
@@ -2349,6 +2402,38 @@
...
+
+  /**
+   * @since 5.5
+   */
+  public static final String EFFORT_TO_REACH_SECURITY_RATING_A_KEY = "effort_to_reach_security_rating_a";
+
+  /**
+   * @since 5.5
+   */
+  public static final Metric<Long> EFFORT_TO_REACH_SECURITY_RATING_A = new Metric.Builder(EFFORT_TO_REACH_SECURITY_RATING_A_KEY, "Effort to reach security rating A", Metric.ValueType.WORK_DUR)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Split this 192 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/e904bc56e18180771d18f7f3fc42d14f9448704c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/e904bc56e18180771d18f7f3fc42d14f9448704c/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L260) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/81d1a131bdfc983ec8d1515c3c8df7e6ec5e7c60/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)
* MAJOR [The "input" parameter nullability is different in the superclass method, and that should not be changed.](https://github.com/SonarSource/sonarqube/blob/81d1a131bdfc983ec8d1515c3c8df7e6ec5e7c60/server/sonar-server/src/main/java/org/sonar/server/computation/qualitymodel/NewQualityModelMeasuresVisitor.java#L260) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by teryk** [comment]
lgtm-monkey

**Pull request merged**